### PR TITLE
add README.md, RustString library, Iterator, general functions

### PR DESCRIPTION
**Pull request by ntson-aws**
README.md: documentation (installation, description) of the library
RustString.lean: the RustString library which includes all the equivalent Lean versions of Rust String functions in Rust standard library, proof of correctness and supporting theorems/lemma
Basic.lean: Generic lemmas for logic, Nat, List
Iterator: modeling Rust Iterator in Lean

### CODE REVIEW

**Review by BrunoDutertre — COMMENTED** [review]
Can you move the Proof_Tutorial & Proof_Example to a separate directory?

Also, we should try to follow the Lean naming conventions and remove the underscore from the file names.

I haven't checked, but there are conventions here: https://leanprover-community.github.io/contribute/style.html

File: README.md [hunk 1]
```diff
@@ -1,7 +1,154 @@
...
+---> 
+
 ## rust-lean-models
 
 Lean models of various Rust libraries to facilitate Lean-based verification of Rust programs.
 
+
+## Description
+The library defines equivalent Lean versions of functions from the Rust standard library.
+This library is intended to be used for verifying string-manipulating Rust programs via Lean.
```
Comment (zhassan-aws) [inline README.md]:
```suggestion
This library is intended to be used for verifying Rust programs via Lean.
```

File: README.md [hunk 2]
```diff
@@ -1,7 +1,154 @@
...
+
+The Lean implementation is based on the description of the Rust functions, which are published at https://doc.rust-lang.org/std.
+The library includes:
+- Definitions of functions equivalent to those from the Rust standard library
+- Proofs that these definitions are correct (i.e., match our understanding of the Rust original)
```
Comment (zhassan-aws) [inline README.md]:
```suggestion
- Proofs that these definitions are consistent with the description of the Rust functions
```

File: README.md [hunk 3]
```diff
@@ -1,7 +1,154 @@
...
+- Supporting lemmas and theorems for proof construction
+
+## Installation
+
+To use the library, add the following lines to your `lakefile.lean`:
+
+` require rust-lean-models from git "https://github.com/model-checking/rust-lean-models" `
```
Comment (zhassan-aws) [inline README.md]: Can you test if this works using your fork?

File: README.md [hunk 4]
```diff
@@ -1,7 +1,154 @@
...
+
+Then do `lake update` in the terminal.
+
+## Usage
+### Translating Rust program
```
Comment (zhassan-aws) [inline README.md]:
```suggestion
### Translating a Rust program 
```
OR
```suggestion
### Translating Rust programs
```

File: README.md [hunk 5]
```diff
@@ -1,7 +1,154 @@
...
+To use the library, add the following lines to your `lakefile.lean`:
+
+` require rust-lean-models from git "https://github.com/model-checking/rust-lean-models" `
+
+Then do `lake update` in the terminal.
+
+## Usage
+### Translating Rust program 
+
+For any Rust built-in functions which are used in user code , map it with
```
Comment (zhassan-aws) [inline README.md]:
```suggestion
For any Rust built-in functions which are used in user code, map it with 
```

File: README.md [hunk 6]
```diff
@@ -1,7 +1,154 @@
...
+Then do `lake update` in the terminal.
+
+## Usage
+### Translating Rust program 
+
+For any Rust built-in functions which are used in user code , map it with 
+the corresponding function name in the library (see the Tables below).
+
+### Proof Tutorial
+We demonstrate some proof techniques in Proof_Tutorial.lean through two simple programs that compute
```
Comment (zhassan-aws) [inline README.md]: Please make Proof_Tutorial.lean a link to the actual file so that people can click on it.

File: README.md [hunk 7]
```diff
@@ -1,7 +1,154 @@
...
+the longest common prefix and longest common substring of the two input strings. 
+More examples can be found in Proof_Example.lean
```
Comment (zhassan-aws) [inline README.md]: Same here

File: README.md [hunk 8]
```diff
@@ -1,7 +1,154 @@
...
+
+## Implementation
+
+
+### Recursive function definition
+For each Rust function, we provide a recursive Lean function. Implementing 
+the equivalent functions in Lean recursively enables user to construct 
+induction proofs conveniently. The Lean function has the same as the Rust original,
```
Comment (zhassan-aws) [inline README.md]: > The Lean function has the same as the Rust original

Missing "name"?

File: README.md [hunk 9]
```diff
@@ -1,7 +1,154 @@
...
+More examples can be found in Proof_Example.lean
+
+## Implementation
+
+
+### Recursive function definition
+For each Rust function, we provide a recursive Lean function. Implementing 
+the equivalent functions in Lean recursively enables user to construct 
+induction proofs conveniently. The Lean function has the same as the Rust original, 
+except when the Rust name  clashes with a Lean keyword. In case of a clash, a Rust function 'func_name'
```
Comment (zhassan-aws) [inline README.md]:
```suggestion
except when the Rust name clashes with a Lean keyword. In case of a clash, a Rust function 'func_name' 
```

File: README.md [hunk 10]
```diff
@@ -1,7 +1,154 @@
...
+induction proofs conveniently. The Lean function has the same as the Rust original, 
+except when the Rust name  clashes with a Lean keyword. In case of a clash, a Rust function 'func_name' 
+is renamed to `rust_func_name` in Lean.
+
+Implementing a function recursively may requires some extra parameters.
+In those cases, first, we implement an auxiliary function (name: `func_name_aux`) which is defined 
+recursively with the parameters, then we define the function `func_name` based on the auxiliary function 
+with initial values for the parameters. 
+For example, `char_indices` is defined based on `char_indices_aux` as 
+`def char_indices (s: Str):= char_indices_aux s 0`.
```
Comment (zhassan-aws) [inline README.md]:
```suggestion
`def char_indices (s: Str) := char_indices_aux s 0`.
```

Comment (zhassan-aws) [reply README.md]: Can you add "where the parameter is ..." and describe what the parameter denotes.

File: README.md [hunk 11]
```diff
@@ -1,7 +1,154 @@
...
+
+For Rust functions involving the `Pattern` type, we implement two recursive sub-functions 
+(name: `func_name_char_filter` and  `func_name_substring`) which replace the `Pattern` type 
+in the input by either a char filter (or type`Char → Bool`) or a `List Char`. Then we define 
+the function `func_name` based on these two sub-functions by matching the input of `Pattern` type.
+For example, `split` is defined by two sub-functions `split_char_filter` and `split_substring` as: 
+
+```
```
Comment (zhassan-aws) [inline README.md]:
```suggestion
```lean
```
Please add `lean` to all code snippets to enable syntax highlighting.

File: README.md [hunk 12]
```diff
@@ -1,7 +1,154 @@
...
+def split (s: Str) (p: Pattern) := match p with
+| Pattern.SingleChar c => split_char_filter s (fun x => x = c)
+| Pattern.ListChar l => split_char_filter s (fun x => x ∈ l)
+| Pattern.FilterFunction f => split_char_filter s f
+| Pattern.WholeString s1 => split_substring s s1
+```
+All recursive implementations are proven to be "correct" in the sense that they are consistent with
+the Rust original (see below for more details).
```
Comment (zhassan-aws) [inline README.md]: I would reword this to "are consistent with the descriptions of the Rust versions".

File: README.md [hunk 13]
```diff
@@ -1,7 +1,154 @@
...
+
+## Correctness Proofs
+
+### For functions that return Bool or can be defined based on other functions that have already been proven correct
```
Comment (zhassan-aws) [inline README.md]:
```suggestion
### For functions that return `bool` or can be defined based on other functions that have already been proven correct
```

File: README.md [hunk 14]
```diff
@@ -1,7 +1,154 @@
...
+   
+- First, we provide a variant Lean definition of the Rust function that we call the definitional 
+version (with name`func_name_def`).  This version is intended to match the documented description
```
Comment (zhassan-aws) [inline README.md]:
```suggestion
version (with name `func_name_def`).  This version is intended to match the documented description 
```

File: README.md [hunk 15]
```diff
@@ -1,7 +1,154 @@
...
+All recursive implementations are proven to be "correct" in the sense that they are consistent with
+the Rust original (see below for more details).
+
+## Correctness Proofs
+
+### For functions that return Bool or can be defined based on other functions that have already been proven correct
+   
+- First, we provide a variant Lean definition of the Rust function that we call the definitional 
+version (with name`func_name_def`).  This version is intended to match the documented description 
+of the Rust function. Ideally, the definitional version should not contains recursion, except in some trivial cases,
```
Comment (zhassan-aws) [inline README.md]:
```suggestion
of the Rust function. Ideally, the definitional version should not contain recursion, except in some trivial cases, 
```

File: README.md [hunk 16]
```diff
@@ -1,7 +1,154 @@
...
+- First, we provide a variant Lean definition of the Rust function that we call the definitional 
+version (with name`func_name_def`).  This version is intended to match the documented description 
+of the Rust function. Ideally, the definitional version should not contains recursion, except in some trivial cases, 
+but it can make use of the recursive versions of other functions that have been previously proven correct.
+
+- Then, we state and prove a theorem that shows that the recursive and definitional versions of Rust 
+function `func_name` are equivalent. This equivalence theorem is called `func_name_EQ` and 
+it has type `func_name = func_name_def`.
+The theorem ensures that the function is implemented correctly 
+and allows the two versions to be used interchangably.
```
Comment (zhassan-aws) [inline README.md]:
```suggestion
and allows the two versions to be used intercheangably. 
```

File: README.md [hunk 17]
```diff
@@ -1,7 +1,154 @@
...
+
+- For example, the function `is_char_boundary` has a definitional version: 
+
+    `def is_char_boundary_def (s: Str) (p: Nat) := p ∈ (ListCharPos s ++ [byteSize s])`
+
+    and an equivalence theorem 
+
+    `theorem is_char_boundary_EQ : is_char_boundary s p =  is_char_boundary_def s p`.
+
+When Rust description cannot be efficiently expressed in Lean (require recursions, or very unintuitive),
```
Comment (zhassan-aws) [inline README.md]:
```suggestion
When the description of a Rust function cannot be efficiently expressed in Lean (requires recursions, or is unintuitive),
```

File: README.md [hunk 18]
```diff
@@ -1,7 +1,154 @@
...
+    then define `split_inclusive_char_filter_def` based on them:
+
+    ```
+    def split_inclusive_char_filter_def (s: Str) (f: Char → Bool):= split_at_charpos_list s (List.map (fun x => x+1) (list_char_filter_charpos s f))
+    ```
+defined based on the
+### When the Rust documentation describes properties of the return value 
+We state and prove a soundness theorem for the function with
+name: `func_name_sound` and type: `x = func_name input1 input2 ...  ↔ properties of x`.
+For example, the soundness thorem for the function `floor_char_boundary` is
```
Comment (zhassan-aws) [inline README.md]:
```suggestion
For example, the soundness theorem for the function `floor_char_boundary` is 
```

**Review by zhassan-aws — COMMENTED** [review]
Some comments on the README. I'm still looking at the other files.

File: README.md (hunk 16 above)
Comment (ntson-aws) [reply README.md]: interchangeably is correct

File: RustLeanModels/Basic.lean [hunk 19]
```diff
@@ -1 +1,544 @@
...
+  rename_i n ind
+  by_cases (∀ m < n, ¬ p m ); rename_i gm
+  . exists n ; simp only [gn, true_and]; intro x gx ;
+    by_contra; rename_i gc; simp only [not_le] at gc;
+    have gm := gm x gc; contradiction
+  . rename_i gm ; simp only [not_forall, not_not, exists_prop] at gm; obtain ⟨m,gm⟩:=gm
+    have ind:= ind m gm.left gm.right ; obtain⟨x,gx⟩:= ind
+    use x;
+
+lemma imp_eq_not_or: (a → b) ↔  (¬ a ∨ b) :=by
```
Comment (zhassan-aws) [inline RustLeanModels/Basic.lean]: Is this used anywhere?

File: RustLeanModels/Iterator.md [hunk 20]
```diff
@@ -0,0 +1,37 @@
+<!---
```
Comment (zhassan-aws) [inline RustLeanModels/Iterator.md:1]: Would it make sense to move the contents of this file to the beginning of `Iterator.lean`?

File: RustLeanModels/Iterator.md [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+| std::iter::Iterator::all |   List.all    | https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.all |
+| std::iter::Iterator::any |   List.any    | https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.any |
+| std::iter::Iterator::chain |   List.append    | https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.chain |
+| std::iter::Iterator::collect |   <em>self    | https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.collect |
+| std::iter::Iterator::count |   List.length    | https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.count |
+| std::iter::Iterator::emunerate |   List.enum    | https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.enumerate |
+| std::iter::Iterator::fold |   List.foldl    | https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.fold |
+| std::iter::Iterator::last |   List.getLast    | https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.last |
+| std::iter::Iterator::map |   List.map    | https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.map |
+| std::iter::Iterator::nth |   List.get?   | https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.fold |
```
Comment (zhassan-aws) [inline RustLeanModels/Iterator.md]: Should this be `#method.nth`?

File: RustLeanModels/RustString.md [hunk 22]
```diff
@@ -0,0 +1,80 @@
+<!---
+-- Copyright Kani Contributors
+-- SPDX-License-Identifier: Apache-2.0 OR MIT
+---> 
+## Rust String 
+
+## Data type conversion
+- String is converted to `List Char` in Lean. This helps constructing induction proofs.
+- The integer types that are used in String indexing are converted to `Nat` in the Lean implementation.
+This assumes that overflow exceptions will not happen. Note that overflow exeptions can only happen
```
Comment (zhassan-aws) [inline RustLeanModels/RustString.md]:
```suggestion
This assumes that overflow exceptions will not happen. Note that overflow exceptions can only happen 
```

File: RustLeanModels/RustString.md [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+## Rust String 
+
+## Data type conversion
+- String is converted to `List Char` in Lean. This helps constructing induction proofs.
+- The integer types that are used in String indexing are converted to `Nat` in the Lean implementation.
+This assumes that overflow exceptions will not happen. Note that overflow exeptions can only happen 
+in Rust programs which use usize for String indexing when the Strings size are GBs.
+- Byte Lists (for UTF8 conversion) are represented as `List Nat` in Lean. Strings are converted from `List Char` to `List Nat` by the function `Str_toUTF8`. 
+  This function ensures that the output is a valid UTF8 string. We use three axioms: `Char_Pos0`, `Char_Diff`, and `Char_Size` which describe
+  the properties of UTF8 encoding (see UTF8Str.lean).
```
Comment (zhassan-aws) [inline RustLeanModels/RustString.md]: Add a link to `UTF8Str.lean`.

File: RustLeanModels/RustString.md [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+- The trait std::str::pattern::Pattern is converted to the inductive type `Pattern` (see RustString.lean).
```
Comment (zhassan-aws) [inline RustLeanModels/RustString.md:15]:
```suggestion
- The trait `std::str::pattern::Pattern` is converted to the inductive type `Pattern` (see RustString.lean).
```

Comment (zhassan-aws) [reply RustLeanModels/RustString.md:15]: Add a link to `RustString.lean`.

File: RustLeanModels/RustString.md [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+| core::str::split_terminator |   split_terminator    | https://doc.rust-lang.org/std/primitive.str.html#method.split_terminator |
+| core::str::split_whitespace |   split_whitespace    | https://doc.rust-lang.org/std/primitive.str.html#method.split_whitespace |
+| core::str::splitn |   splitn    | https://doc.rust-lang.org/std/primitive.str.html#method.splitn |
+| core::str::starts_with |   starts_with    | https://doc.rust-lang.org/std/primitive.str.html#method.starts_with |
+| core::str::strip_prefix |   strip_prefix    | https://doc.rust-lang.org/std/primitive.str.html#method.strip_prefix |
+| core::str::strip_suffix |   strip_suffix    | https://doc.rust-lang.org/std/primitive.str.html#method.strip_suffix |
+| core::str::to_ascii_lowercase |    to_ascii_lowercase    | https://doc.rust-lang.org/std/primitive.str.html#method.to_ascii_lowercase |
+| core::str::to_ascii_uppercase |   to_ascii_uppercase    | https://doc.rust-lang.org/std/primitive.str.html#method.to_ascii_uppercase |
+| core::str::to_lowercase |   not yet   | https://doc.rust-lang.org/std/primitive.str.html#method.to_lowercase |
```
Comment (zhassan-aws) [inline RustLeanModels/RustString.md]: Change "not yet" to N/A

File: RustLeanModels/RustString.md [hunk 26]
```diff
@@ -0,0 +1,80 @@
...
+| core::str::to_uppercase |   not yet    | https://doc.rust-lang.org/std/primitive.str.html#method.to_uppercase |
```
Comment (zhassan-aws) [inline RustLeanModels/RustString.md]: Same here

File: RustLeanModels/RustString.md [hunk 27]
```diff
@@ -0,0 +1,80 @@
...
+## Implemented functions in Lean library for List
+| Rust String function                 | Lean equivalent function       | Description link |
+| ----------------------------- | ------------------- | ---------------------- |
+| core::str::chars |   <em>self    | https://doc.rust-lang.org/std/primitive.str.html#method.chars |
+| core::str::is_empty |   List.isEmpty    | https://doc.rust-lang.org/std/primitive.str.html#method.is_empty |
+| core::str::len |   List.length    | https://doc.rust-lang.org/std/primitive.str.html#method.len |
+
+## Limitations
+The RustString library does not include:
+- Functions that mutate their input (e.g., make_ascii_lowercase), but it includes the cloning version (e.g. to_ascii_lowercase).
```
Comment (zhassan-aws) [inline RustLeanModels/RustString.md]:
```suggestion
- Functions that mutate their input (e.g., `make_ascii_lowercase`), but it includes the cloning version (e.g. `to_ascii_lowercase`).
```

File: RustLeanModels/RustString.md [hunk 28]
```diff
@@ -0,0 +1,80 @@
...
+- Functions which may panic (e.g., slice indexing), but it includes the non-panicking version (e.g. str::get)
```
Comment (zhassan-aws) [inline RustLeanModels/RustString.md]:
```suggestion
- Functions which may panic (e.g., slice indexing), but it includes the non-panicking version (e.g. `str::get`)
```

File: RustLeanModels/RustString.lean [hunk 29]
```diff
@@ -0,0 +1,4804 @@
...
+General nat: m n
+List nat: l
+Element of list: x
+Extra parameter in aux functions: k
+-/
+
+namespace RustString
+
+
+def rust_isWhitespace (c : Char) := (isWhitespace c) || (c.val = 12)
```
Comment (zhassan-aws) [inline RustLeanModels/RustString.lean]: How did you come across the difference between the Rust and Lean versions of what is considered a whitespace character? Are there values other than 12 on which they disagree on?

It seems they also disagree on 11 as:
```lean
#eval Char.isWhitespace (Char.ofNat 11)
```
returns `false`, but the equivalent Rust returns `true`:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e33f093a0030b74eae907a6f1100da26

File: README.md (hunk 3 above)
Comment (ntson-aws) [reply README.md]: done

**Review by zhassan-aws — COMMENTED** [review]
A few more comments.

File: README.md [hunk 30]
```diff
@@ -20,11 +20,23 @@ The library includes:
...
 ## Installation
 
-To use the library, add the following lines to your `lakefile.lean`:
+-  Add the following lines to your `lakefile.lean`:
 
-` require rust-lean-models from git "https://github.com/model-checking/rust-lean-models" `
+    ` require «rust-lean-models» from git "https://github.com/model-checking/rust-lean-models" `
 
-Then do `lake update` in the terminal.
+- Change the lean version in `lean-toolchain.lean`:
```
Comment (zhassan-aws) [inline README.md]: We'd like to keep the Lean toolchain version up-to-date, so to avoid updating the README.md every time we update the toolchain version, let's just say "Change the Lean version in `lean-toolchain` to match the version used in this package" or something like that.

File: README.md [hunk 31]
```diff
@@ -20,11 +20,23 @@ The library includes:
...
+    ` leanprover/lean4:v4.10.0-rc2 `
+
+- Run `lake update` in the terminal.
+
+- Import the packages and open the namespaces"
+
+    ```
```
Comment (zhassan-aws) [inline README.md]:
````
```lean
````

File: README.md [hunk 32]
```diff
@@ -20,11 +20,23 @@ The library includes:
...
-` require rust-lean-models from git "https://github.com/model-checking/rust-lean-models" `
+    ` require «rust-lean-models» from git "https://github.com/model-checking/rust-lean-models" `
 
-Then do `lake update` in the terminal.
+- Change the lean version in `lean-toolchain.lean`:
+    ` leanprover/lean4:v4.10.0-rc2 `
+
+- Run `lake update` in the terminal.
+
+- Import the packages and open the namespaces"
```
Comment (zhassan-aws) [inline README.md]: Add "in your Lean files" at the end.

File: RustLeanModels/Iterator.md (hunk 21 above)
Comment (ntson-aws) [reply RustLeanModels/Iterator.md]: Yes, it a mistake

File: RustLeanModels/RustString.lean (hunk 29 above)
Comment (ntson-aws) [reply RustLeanModels/RustString.lean]: We should compare the Lean function isWhitespace with the Rust function: is_ascii_whitespace. 
The Rust function is_whitespace is for unicode. I have the Lean version of it in the library: is_Unicode_Whitespace.

File: README.md [hunk 33]
```diff
@@ -1,7 +1,165 @@
...
+- Proofs that these definitions are consistent with the description of the Rust functions
+- Supporting lemmas and theorems for proof construction
+
+## Installation
+
+-  Add the following lines to your `lakefile.lean`:
+
+    ` require «rust-lean-models» from git "https://github.com/model-checking/rust-lean-models" `
+
+- Change the lean version in `lean-toolchain.lean` to that of `"https://github.com/model-checking/rust-lean-models/lean-toolchain.lean"`
```
Comment (zhassan-aws) [inline README.md]:
```suggestion
- Change the lean version in your `lean-toolchain` to the one in https://github.com/model-checking/rust-lean-models/blob/main/lean-toolchain
```

File: RustLeanModels/RustString.lean [hunk 34]
```diff
@@ -0,0 +1,4914 @@
...
+  rename_i h t ind
+  simp only [foldl_cons, add_eq, zero_add] ; rw[@ind h, @ind (x+h)]; omega
+
+lemma sum_List_Nat_cons : sum_list_Nat (h::t) = h + sum_list_Nat t  :=by
+  simp only [sum_list_Nat, foldl_cons, add_eq, zero_add]
+  exact fold_add_para1_elim
+
+
+@[simp]
+def byteSize (s : List Char) : Nat := match s with
```
Comment (zhassan-aws) [inline RustLeanModels/RustString.lean:77]: It's useful to add a description of those functions in the form of doc comments, e.g.
```lean
/--
Compute the length of `s` in terms of the number of bytes.
The length is the sum of the number of bytes for each individual character.
-/

File: RustLeanModels/RustString.lean [hunk 35]
```diff
@@ -0,0 +1,4914 @@
...
+  intro; omega
+
+def ListCharPos_aux (s: Str) (i: Nat):= match s with
+  | [] => []
+  | h::t => match t with
+            | [] => [i]
+            | _::_ => i:: ListCharPos_aux t (i+ Char.utf8Size h)
+
+@[simp]
+def ListCharPos (s: Str) := ListCharPos_aux s 0
```
Comment (zhassan-aws) [inline RustLeanModels/RustString.lean:157]: Same here: add a doc comment that describes what this function is doing.

File: RustLeanModels/RustString.lean [hunk 36]
```diff
@@ -0,0 +1,4914 @@
...
+  simp only [ListCharPos, ListCharPos_aux_sound, ne_eq, add_zero]
+
+
+@[simp]
+def CharBoundPos_aux (l: Str) (s: Nat): List Nat := match l with
+  | [] => [s]
+  | h::t => s :: CharBoundPos_aux t (s + Char.utf8Size h)
+
+@[simp]
+def CharBoundPos (s: Str) : List Nat := CharBoundPos_aux s 0
```
Comment (zhassan-aws) [inline RustLeanModels/RustString.lean:204]: Same here.

File: RustLeanModels/RustString.lean [hunk 37]
```diff
@@ -0,0 +1,4914 @@
...
+
+/-
+Function: is_char_boundary
+From: core::str::is_char_boundary
+Description: https://doc.rust-lang.org/std/primitive.str.html#method.is_char_boundary
+-/
+
+@[simp]
+def is_char_boundary (s: Str) (i: Nat) := match s with
+  | [] => if i = 0 then true else false
```
Comment (zhassan-aws) [inline RustLeanModels/RustString.lean]: Replace with `i == 0` as discussed.

File: RustLeanModels/RustString.lean [hunk 38]
```diff
@@ -0,0 +1,4914 @@
...
+  rename_i hs1 ts1; simp only [byteSize_cons, cons_append, cons.injEq] at g;
+  use ts1; simp only [g, singleton_append, true_and]; rw[← g.right.right]
+  apply ind.mpr; constructor; omega; exact g.right.right
+
+
+lemma split_at_zero: split_at s 0 = some ([], s) := by
+  unfold split_at split_at_aux; simp only [↓reduceIte]
+  split; simp only; simp only
+
+lemma split_at_is_char_boundary_none: ¬ is_char_boundary s i ↔  split_at s i = none :=by
```
Comment (zhassan-aws) [inline RustLeanModels/RustString.lean]: Would it make sense to name the lemma `split_at_non_char_boundary_none` instead?

File: RustLeanModels/RustString.lean [hunk 39]
```diff
@@ -0,0 +1,4914 @@
...
+
+
+--#eval ListCharPos "L∃∀N".data
+--#eval split_at "La∃∀N".data 5
+
+
+/-
+Function: string_slice
+From: core::str::get
+Description: https://doc.rust-lang.org/std/primitive.str.html#method.floor_char_boundary
```
Comment (zhassan-aws) [inline RustLeanModels/RustString.lean]: Should this be https://doc.rust-lang.org/std/primitive.str.html#method.get?

File: RustLeanModels/RustString.lean [hunk 40]
```diff
@@ -0,0 +1,4914 @@
...
+                                                                      some s21
+
+--#eval ListCharPos "La∃∀N".data
+--#eval string_slice "La∃∀N".data  5 1
+
+/-
+For special cases: get(0..a)
+Function: PrefixFromPos
+From: core::str::get
+Description: https://doc.rust-lang.org/std/primitive.str.html#method.floor_char_boundary
```
Comment (zhassan-aws) [inline RustLeanModels/RustString.lean]: Same here

**Comment by BrunoDutertre** [comment]
Another small thing to do: use `lakefile.toml` instead of `lakefile.lean`

The `toml` format is nicer and easier to maintain. You can do
```
lake translate-config toml
```
to build it automatically.

**Review by zhassan-aws — COMMENTED** [review]
A few more comments.

**Review by BrunoDutertre — COMMENTED** [review]
Looks good to me.

File: RustLeanModels/RustString.lean [hunk 41]
```diff
@@ -0,0 +1,4922 @@
...
+  split; simp only [IsEmpty.forall_iff];
+  split; simp_all only [Bool.not_eq_true, not_false_eq_true, not_sym, isEmpty_Prop, not_le,
+    IsEmpty.forall_iff, and_self, forall_true_left]
+  simp_all only [Bool.not_eq_true, not_lt, not_false_eq_true, not_sym, forall_true_left, true_and]
+  exact split_at_aux_none.mp
+
+
+
+
+--#eval ListCharPos "L∃∀N".data
```
Comment (zhassan-aws) [inline RustLeanModels/RustString.lean:644]: It's sometimes useful to add these as examples, e.g.:
```lean
example : ListCharPos "L∃∀N".data = [0, 1, 4, 7] := by rfl
```

File: RustLeanModels/UTF8Str.lean [hunk 42]
```diff
@@ -0,0 +1,150 @@
...
+open Nat
+set_option maxHeartbeats 10000000
+
+
+--TODO: prove the axiom from utf8EncodeChar function in Data/String/Extra.lean
+opaque toByte : (Char ×  Nat) →  Nat
+
+def isValid (e : Char ×  Nat) : Bool := e.2 < utf8Size e.1
+
+axiom Char_Pos0 : ∀ c1 c2 b, b ≠ 0 → toByte (c1,0) ≠ toByte (c2,b)
```
Comment (zhassan-aws) [inline RustLeanModels/UTF8Str.lean]: Please add a comment that describes what this axiom is stating.

Also, don't we need `isValid (c2, b)` as a hypothesis?

File: RustLeanModels/UTF8Str.lean [hunk 43]
```diff
@@ -0,0 +1,150 @@
...
+
+axiom Char_Diff : ∀ c1 c2, c1 ≠ c2 → ∃ b , b < utf8Size c1 ∧  toByte (c1,b) ≠ toByte (c2,b)
```
Comment (zhassan-aws) [inline RustLeanModels/UTF8Str.lean]: I would expect this to state:
```lean
∀ c1 c2, c1 ≠ c2 → (utf8Size c1 ≠ utf8Size c2) ∨ (∃ b, b < utf8Size c1 ∧  toByte (c1,b) ≠ toByte (c2,b))
```

File: RustLeanModels/UTF8Str.lean [hunk 44]
```diff
@@ -0,0 +1,150 @@
...
+
+axiom Char_Size : ∀ c1 c2,  toByte (c1,0) = toByte (c2,0) →  utf8Size c1 = utf8Size c2
```
Comment (zhassan-aws) [inline RustLeanModels/UTF8Str.lean]: Can you cite a reference for this axiom?

File: RustLeanModels/UTF8Str.lean [hunk 45]
```diff
@@ -0,0 +1,150 @@
...
+
+def Char_toUTF8 (c : Char) (i : Nat) : List Nat := match i with
```
Comment (zhassan-aws) [inline RustLeanModels/UTF8Str.lean]: Please add a comment that describes what this is doing.

**Review by zhassan-aws — COMMENTED** [review]
Can you make the PR description more elaborate?

**Review by zhassan-aws — APPROVED** [review]
